File: src/main/java/org/opendsb/messaging/BaseMessage.java
package org.opendsb.messaging;

import java.util.UUID;

import org.opendsb.json.AnnotationExclusion;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

public abstract class BaseMessage implements Message {

	private String messageId = UUID.randomUUID().toString();

	protected MessageType type;

	private String origin;

	private String destination;

	private String latestHop;

	public BaseMessage(String destination, String origin) {
		super();
		this.origin = origin;
		this.latestHop = origin;
		this.destination = destination;
	}

	public BaseMessage(String origin) {
		super();
		this.origin = origin;
		this.latestHop = origin;
	}

	@Override
	public String getMessageId() {
		return messageId;
	}

	@Override
	public String getDestination() {
		return destination;
	}

	@Override
	public MessageType getType() {
		return type;
	}

	@Override
	public String getOrigin() {
		return origin;
	}

	@Override
	public String getLatestHop() {
		return latestHop;
	}

	@Override
	public void setLatestHop(String latestHop) {
		this.latestHop = latestHop;
	}

	@Override
	public String toJSON() {
		Gson gson = new GsonBuilder().setExclusionStrategies(new AnnotationExclusion()).create();
		return gson.toJson(this);
	}

	@Override
	public String toString() {
		return "BaseMessage [messageId=" + messageId + ", type=" + type + ", destination=" + destination + "]";
	}
}

File: python/opendsb/src/opendsb/messaging/basemessage.py
import uuid

from opendsb.messaging.message import Message, MessageType


class BaseMessage(Message):
    def __init__(
        self, origin: str, destination: str, type: MessageType, id: str | None = None, latest_hop: str | None = None
    ):
        self.origin = origin
        self.destination = destination
        self.type = type
        self.id = id or str(uuid.uuid4())
        self.latest_hop = latest_hop or origin

    def __repr__(self) -> str:
        return f"{self.__class__.__qualname__}(origin={self.origin}, destination={self.destination}, type={self.type}, id={self.id})"
